src/Workspaces/Core/Portable/Workspace/Solution/ProjectChanges.cs

```diff
@@ -191,5 +216,16 @@ public IEnumerable<DocumentId> GetRemovedAdditionalDocuments()
                 }
             }
         }
+
+        public IEnumerable<DocumentId> GetRemovedAnalyzerConfigDocuments()
+        {
+            foreach (var doc in _oldProject.AnalyzerConfigDocuments)
+            {
+                if (!_newProject.ContainsAnalyzerConfigDocument(doc.Id))
+                {
+                    yield return doc.Id;
+                }
+            }
+        }
```

Comment:
@jasonmalinowski How opposed would you be to a PR that tried to unify all this duplication?  i'm potentially interested in writing it, but i'd prefer to only do it if i had reasonable confidence it would be taken.

Reply:
I am +1 on a follow-up PR to encapsulate all the document handling in a central way, without us having to handle all document kinds at every place in different layers, and then have helpers with bunch of lambdas to share code. I will still try to do the later for large methods that I have cloned in this PR, but still like the overall idea. It would be super beneficial if we add a new document kind in future.

Reply:
Yes, I'm all for cleaning this up, at least in principle. I think a lot of this can be done just by extracting out of ProjectState a "abstract collection of documents" thingy of some sort that has lots of these common operations, and so then we're working with those internally. I'm not sure if that's something we'd expose or still just have the (current) public methods just forward to it; I'm actually OK with that being internal implementation since the forwarding methods would be trivial.

If you have a concrete pattern in your head it might be worth discussing prior doing it so we can try to forsee gotchas, but I fully confess if it was me trying to do the dedup I'd just dive in head first because you'll learn things much faster that way.

Reply:
That's pretty much exactly waht i was thinking.  On our public surface area, we'd still expose the split.  But all those public methods would just callinto the raw instance of the specific collection which had all the shared logic.  So you'd basically have:

```C#
class Solution {
    _normalDocuments = new NormalDocumentCollection();
    _analyzerDocuments = new AnalyzerDocumentCollection();
    _configDocuments = new ConfigDocumentCollectoin();
}

class AbstractDocumentCollection<TDocument> where TDocument : TextDocument {
    // all our algorithms and query/mutation methods.
    abstract void Something();
}

class ConfigDocumentCollection : AbstractDocumentCollection<ConfigDocument> {
    // overrides to make tings specific to config docs.
}
```

The leaf DocCollections would be extremely simple, and we'd put all shareable logic in one place.